compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/AIETarget.cpp

```diff
@@ -38,6 +38,43 @@ static llvm::cl::opt<std::string> clEnableAMDAIEUkernels(
                    "unprefixed microkernels to enable, e.g. `matmul`."),
     llvm::cl::init("none"));
 
+// Utility to find aie.device Op corresponding to the export Op, We have
+// hal.executable.variant {
+//   hal.executable.export symbol1
+//   hal.executable.export symbol2
+//   module {
+//     aie.device {
+//       ...
+//       func.func symbol1
+//     }
+//     aie.device {
+//       ...
+//       func.func symbol2
+//     }
+//   }
+// }
+// Hence we need to find the func.func that coresponds to the export op symbol
+// and return its parent aie.device Op. This is what we will pass to the
+// `aie2xclbin` tool for artifact generation per entry point
+static xilinx::AIE::DeviceOp getDeviceOpFromEntryPoint(ModuleOp moduleOp,
+                                                       StringRef exportOpName) {
+  xilinx::AIE::DeviceOp deviceOp;
+
+  moduleOp.walk([&](func::FuncOp funcOp) {
+    if (funcOp.getSymName() == exportOpName) {
+      deviceOp = dyn_cast_or_null<xilinx::AIE::DeviceOp>(funcOp->getParentOp());
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  if (!deviceOp) {
+    moduleOp.emitError()
+        << "failed to find aie.device Op containing func.func with symbol "
```

Comment:
nit: why `aie.device Op` but `func.func` -- they're both MLIR Ops. I'd personally drop the ` Op`.